CAM: Integration of the tool and shape editors from BTL (Better Tool Library)

knipknap opened the pull request:
This very large PR essentially replaces CAM Tool Management completely, as part of the [Better Tool Library](https://github.com/knipknap/better-tool-library/) integration.

## Relevant Issues

Discussion: #9466

I also went through all open CAM issues to find any that may be resolved or otherwise affected:

- Resolves: #8201
- Resolves: #8526
- Resolves: #9625
- Resolves: #11522
- Resolves: #12712
- Likely resolves: #12823, but should be tested to be sure
- Possibly resolves #13603, to be tested
- Resolves: #17083
- Possibly makes obsolete #18598, though the long discussion there seems to diverge a bit from the original problem, so not sure.
- Possibly resolves #18887, needs testing
- Resolves: #20187
- Resolves #20880  (though already closed because it was caused by BTL, not by CAM)

# Release Notes

# Replacing FreeCAD CAM Tool Management

This PR for FreeCAD's CAM workbench marks a significant evolution in how tool bits and their associated assets are managed. This extensive development effort introduces not only a powerful new backend architecture but also crucial user interface enhancements that streamline the CAM workflow and improve usability. The changes prioritize a more modular, flexible, and user-friendly system for handling tool libraries, tool bits, and shapes.

## Enhanced User Experience: Key UI Improvements

A primary focus of the `btl` branch was to improve the user interface for managing tools. Several key user interface components have been introduced or changed to provide a more intuitive and efficient workflow.

### The New ToolBit Editor

The ToolBit Editor has been completely replaced. The new UI provides users with a clear visual representation of the tool, and allows to modify tool all properties - previously only dimensions were editable, now users can also edit mutiple selection fields (enums) and switches (bool).

Another improvement is the ability to edit ToolBits while they are detached from the main document, so the editor can now be opened without needing to open a document first.

![image](https://github.com/user-attachments/assets/9a33f0a3-9080-4210-9a40-1cb1d19c4afa)


### The New ToolBit Selector

Replacing the older, less intuitive methods, the new ToolBit Selector provides a streamlined interface for choosing tool bits. This selector is designed to make it easier for users to browse, search, and select from their available tool libraries and individual tool bits. It also includes a toolbit-specific summary, offering quick, relevant information about each tool to aid in selection.

![image](https://github.com/user-attachments/assets/cd1fc08d-fb2d-46f8-bac0-c8f5b8c9d60f)

The dock was also replaced and features a search field and sorting:

![image](https://github.com/user-attachments/assets/0f44085f-4ae9-43a9-9ebc-aebbb86f131b)


### The New Shape Selector

Complementing the ToolBit Selector, a new, user-friendly Shape Selector has been introduced specifically for choosing tool shapes. This replaces the old file selector approach, making it much simpler to find and assign the correct geometric shape to a tool bit from the available shape assets.

Before:
![image](https://github.com/user-attachments/assets/2236c557-b445-4c36-be8e-85efe34b0331)

After:
![image](https://github.com/user-attachments/assets/1795c1c4-eeba-4e2e-a6fe-c5566cc5bef2)

Furthermore, toolbit shape parameters are now directly editable within the property panel if the tool shape exists locally, with changes taking immediate effect. This eliminates previous limitations and speeds up the process of fine-tuning tools.

![image](https://github.com/user-attachments/assets/349ffc48-3dc5-40fa-adb5-39e2879a7ea4)


### Import/Export Dialogs

New dedicated dialogs have been added to improve the process of managing assets. These include the `AssetOpenDialog` and `AssetSaveDialog`, as well as a specific `ToolBitOpenDialog`. These dialogs streamline the process of importing and exporting tool bits and other assets, providing a more consistent and user-friendly experience.

These significant UI enhancements collectively contribute to a more polished and efficient user experience when working with CAM tools in FreeCAD.

## A Robust Backend: Decoupling and Componentization

Underpinning the improved user interface is a substantial refactoring of the CAM tool management backend. The core problem addressed was the tight coupling between CAM objects and their storage mechanisms, which made the system rigid and difficult to extend. This PR introduces a more modular and decoupled architecture with several key features:

### ToolBits Outside Documents

A significant architectural change allows `ToolBit` instances to exist and be manipulated independently of an active FreeCAD document. This refactoring is crucial for improving performance and enabling more flexible asset management workflows in the future.

### Separation of Concerns

The refactoring effort focused on clearly separating different aspects of tool management:

*   **View:** Handled by the UI components, responsible solely for presentation and user interaction.
*   **Object Model:** The in-memory representation of assets like `ToolBit`, `ToolBitShape`, and `Library`. These objects now implement a common interface and delegate persistence to other components.
*   **Serialization:** The process of converting objects to and from bytes is handled by dedicated serializers, defined by the `AssetSerializer` protocol. This allows for supporting various file formats (native, Camotics, LinuxCNC) through different serializer implementations and enables import and export of new formats in the future.
*   **Storage:** The persistence of assets to various backends (like the file system or potentially online databases) is managed by the new Asset Manager.

### Componentization

The system is broken down into more granular, specialized components. This includes:

*   **Explicit `ToolBitShape` Class:** A dedicated class for tool bit shapes as assets, allowing them to be managed independently and defining their properties in a structured way. **Shape schemas** are introduced to ensure data integrity for shape files.
*   **`ToolBit` Subclasses:** Introduction of specific subclasses such as `ToolBitEndmill` and `ToolBitChamfer`. This allows CAM operations to rely on specific attributes for different tool types, improving code clarity and maintainability.

### The Role of the Asset Manager

Within this new architecture, the generic **AssetManager** plays a crucial role in managing the storage and retrieval of assets. It provides a unified interface for different storage backends (`FileStore`, `MemoryStore`) and handles dependencies between assets. The Asset Manager is designed to support asynchronous operations, which is a key enabler for future features like integrating with online asset databases without blocking the FreeCAD UI.
Assets in the CAM workbench may include:

*   Tool bit libraries
*   Tool bits
*   Tool bit shapes
*   Tool bit shape icons
*   Machines (future)
*   Fixtures (future)
*   Post processors (future)

The building blocks of the Asset Manager are illustrated in the following diagram:

```mermaid
classDiagram
    direction LR

    class AssetManager
    class AssetStore
    class FileStore
    class MemoryStore
    class AssetSerializer
    class Asset
    class ToolBit
    class ToolBitShape
    class Library
    class Machine

    AssetStore *-- AssetManager: has many
    FileStore --|> AssetStore: is
    MemoryStore --|> AssetStore: is
    AssetSerializer *-- AssetManager: has many
    Asset --> AssetSerializer: uses
    Asset *-- AssetManager: creates

   ToolBit --|> Asset: is
   ToolBitShape --|> Asset: is
   Library --|> Asset: is
   Machine --|> Asset: is
```

This diagram highlights the central role of the AssetManager and the clear separation between assets (like `ToolBit` and `ToolBitShape`), storage backends (`FileStore`, `MemoryStore`), and serializers (`AssetSerializer`).

More in-depth explanation can be found in the [AssetManager README](https://github.com/knipknap/FreeCAD/blob/btl/src/Mod/CAM/Path/Tool/assets/README.md).

## Backward Compatibility and Migration

Crucially, the PR maintains backward compatibility as follows:

- Tool library files (.fctl) **remain fully compatible** and are unchanged.
- ToolBit files (.fctb) **remain fully compatible**.
- Existing .FCStd files already containing ToolBits **remain fully compatible** due to a migration function that automatically upgrades them according to the latest changes (e.g. adding ToolID and ShapeID attributes) when the document is opened, ensuring a smooth transition for users.
- Some custom TooBitShape files (.fcstd) **may require an update**: The name of the body in the shape file is used to identify the shape type. I.e. it must be one of the following: 'Ballend', 'Chamfer', 'Dovetail', 'Drill', 'Endmill', 'Fillet', 'Probe', 'Reamer', 'SlittingSaw', 'Tap', 'ThreadMill', 'Bullnose', 'VBit'. Without this update, FreeCAD will not be able to identify the type of shape, and will fall back to "Endmill" if it cannot make a better guess. A warning with instructions is generated if such files are encountered.

## Comprehensive Testing

The branch includes a significant number of new tests to ensure the stability and correctness of the new architecture and UI components. This extensive test suite was vital in identifying and resolving issues, particularly concerning cross-platform compatibility.

## Future Potential

The new asset management system lays a strong foundation for future enhancements. The architecture is designed to support integration with online databases and Git repositories, opening up possibilities for sharing and accessing a wider range of CAM assets directly within FreeCAD.

## Conclusion

The PR represents a substantial and highly beneficial overhaul of the FreeCAD CAM workbench's tool management system. By prioritizing user interface improvements like the new ToolBit Editor and Selector, and implementing a decoupled, componentized backend architecture centered around the Asset Manager, the branch delivers a more usable, flexible, and maintainable system. These changes not only address existing limitations but also pave the way for exciting future developments in CAM asset management within FreeCAD.

luzpaz commented on src/Mod/CAM/CAMTests/PathTestUtils.py:
```diff
@@ -196,3 +206,43 @@ def assertSuccessfulRecompute(self, doc, *objs, msg=None):
+
+class PathTestWithAssets(PathTestBase):
+    """
+    A base class that create an AssetManager, so test can easily fetch
```
`creates`

knipknap commented:
Thank you again for testing, and for the test file. I will fix it of course.

By rebasing, do you mean squashing commits? I am asking because rebasing doesn't reduce the number of commits.

I think realistically there is no way to group the commits in a way that would make sense to review the individual diffs - there are many refactorings included on the way and in the end, almost nothing is left of the original code.

There are probably just a hundred lines of code or so left from the Mod/CAM/Path/Tool folder currently in master. I think the only way to realistically be able to review it is to treat it as an entirely new module.

sliptonic commented:
Usually when I see 100s of commits it is because the author merged upstream into their branch so the commits are duplicates.  It looks like you commit often which is fine.  We would still like to be kind to the freecad history so I would suggest an interactive rebase to squash it down to a manageable number of useful commits.

knipknap commented:
@sliptonic Alright! Both issues you have see are fixed, and the commits are now squashed.

jffmichi commented:
When I'm opening the ToolBit Dock or the ToolBit Library Editor I'm getting a lot of errors and warnings for my tools, e.g. the following:
```
21:55:11  base.ERROR: failed to infer shape type from drillshank; using "endmill"
21:55:11  base.WARNING:  ToolBit drill_1.0mm Parameter 'TipAngle' not found on ToolBitEndmill (Endmill(CuttingEdgeHeight=10,0 mm, Diameter=1,0 mm, Flutes=0, Length=38,0 mm, ShankDiameter=3,17 mm)) 'Drill 1.0mm'. Skipping.
21:55:11  base.ERROR: failed to infer shape type from drillshank; using "endmill"
21:55:11  base.WARNING:  ToolBit drill_1.2mm Parameter 'TipAngle' not found on ToolBitEndmill (Endmill(CuttingEdgeHeight=7,0 mm, Diameter=1,2 mm, Flutes=0, Length=38,0 mm, ShankDiameter=3,17 mm)) 'Drill 1.2mm'. Skipping.
21:55:11  base.WARNING: ToolBit end_mill_6.0mm_1f Attribute 'btl-stickout' not found on ToolBitEndmill (Endmill(CuttingEdgeHeight=22,0 mm, Diameter=6,0 mm, Flutes=1, Length=50,0 mm, ShankDiameter=6,0 mm)) 'End Mill 6.0mm 1F'. Skipping.
21:55:11  base.WARNING:  ToolBit spotting_drill_6.0mm_120deg Parameter 'TipDiameter' not found on ToolBitChamfer (Chamfer(CuttingEdgeAngle=120.000 °, CuttingEdgeHeight=1.732 mm, Diameter=6.000 mm, Flutes=0, Length=50.000 mm, ShankDiameter=6.000 mm)) 'Spotting Drill 6.0mm 120°'. Skipping.
```

It doesn't seem to work with my custom tool shapes, i.e. drillsank.

The menu button CAM -> ToolBit Library Editor still opens the old tool library.

Double clicking any tool gives the following error in the console and a message box:

```
22:00:04  Error instantiating asset 'toolbit://end_mill_6.0mm_1f' of type 'ToolBit': Dependency for shape 'endmill.fcstd' not found by uri toolbitshape://endmill {AssetUri('toolbitshape://endmill'): None}
Traceback (most recent call last):
  File "/usr/lib/freecad/Mod/CAM/Path/Tool/assets/manager.py", line 238, in _build_asset_tree_from_data_sync
    final_asset = asset_class.from_bytes(
        construction_data.raw_data,
    ...<2 lines>...
        serializer,
    )
  File "/usr/lib/freecad/Mod/CAM/Path/Tool/assets/asset.py", line 87, in from_bytes
    return serializer.deserialize(data, id, dependencies)
           ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/freecad/Mod/CAM/Path/Tool/toolbit/serializers/fctb.py", line 87, in deserialize
    raise ValueError(
        f"Dependency for shape '{shape_id}' not found by uri {shape_uri}" f" {dependencies}"
    )
ValueError: Dependency for shape 'endmill.fcstd' not found by uri toolbitshape://endmill {AssetUri('toolbitshape://endmill'): None}
22:00:04  editor.ERROR: Failed to load or edit toolbit asset toolbit://end_mill_6.0mm_1f: 'NoneType' object has no attribute '_tool_bit_shape'
22:00:05  Traceback (most recent call last):
22:00:05    File "/usr/lib/freecad/Mod/CAM/Path/Tool/library/ui/editor.py", line 445, in toolEdit
    editor_dialog = ToolBitEditor(bit, self.form)  # Create dialog instance
22:00:05    File "/usr/lib/freecad/Mod/CAM/Path/Tool/toolbit/ui/editor.py", line 225, in __init__
    widget = ShapeWidget(toolbit._tool_bit_shape)
                         ^^^^^^^^^^^^^^^^^^^^^^^
22:00:05  AttributeError: 'NoneType' object has no attribute '_tool_bit_shape'
22:00:05  
```

Clicking "Add Existing" gives the following error and nothing happens:

```
22:01:52  RecursionError: maximum recursion depth exceeded
```

Clicking "Create Toolbit" opens an empty window:

![grafik](https://github.com/user-attachments/assets/e5331b2f-e590-43c3-adf9-4a0c25ade167)
![grafik](https://github.com/user-attachments/assets/849a73aa-23d2-4c5e-bcfe-1858c5921318)

```
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/freecad/Mod/CAM/Path/Tool/assets/manager.py", line 238, in _build_asset_tree_from_data_sync
    final_asset = asset_class.from_bytes(
        construction_data.raw_data,
    ...<2 lines>...
        serializer,
    )
  File "/usr/lib/freecad/Mod/CAM/Path/Tool/shape/models/base.py", line 279, in from_bytes
    shape_class = ToolBitShape.get_shape_class_from_bytes(data)
  File "/usr/lib/freecad/Mod/CAM/Path/Tool/shape/models/base.py", line 177, in get_shape_class_from_bytes
    raise ValueError(f"Error processing FCStd data: {e}")
ValueError: Error processing FCStd data: No ToolBitShape subclass found matching Body label 'Body'
```

Clicking the save button gives the following error without any other effect:

```
22:05:23  TypeError: AssetSaveDialog.exec() missing 1 required positional argument: 'asset'
22:05:23  
```

knipknap commented:
Phew, that's a lot of issues. Can you attach your custom tool shape for testing, please? I'll look into the other issues.

Thank you for testing!

jffmichi commented:
[Tools.zip](https://github.com/user-attachments/files/20303299/Tools.zip)

jffmichi commented:
Do I understand correctly that this editor is not yet part of the PR?
https://github.com/knipknap/better-tool-library/raw/main/media/library.png

knipknap commented:
> Do I understand correctly that this editor is not yet part of the PR? https://github.com/knipknap/better-tool-library/raw/main/media/library.png

Yes, that is correct. Though the Dock now comes so close to the tool editor look and feel that I wonder if keeping the editor is even the best way to go. I could just add the add/remove/import/export functions to the dock instead. But that is a discussion I wanted to start after this PR :smile: , for now I did not want to include it, yet.

jffmichi commented:
Ah ok good to know. My bad on that point then 😀
Old library editor is fine with me but should be fixed ofc.

knipknap commented:
So, findings and fixes applied:

### Warning "failed to infer shape type" and "No ToolBitShape subclass found matching Body label 'Body'"

My theory above was correct. I improved two things to handle this better. 1. shape inference tries to "guess better" as a fallback, instead of just using the endmill. 2. I added a better error message that also explains how to fix the shape file:

`base.WARNING: failed to infer shape type from centerdrill, guessing "Drill". To fix, name the body in the shape file to one of: ['Ballend', 'Chamfer', 'Dovetail', 'Drill', 'Endmill', 'Fillet', 'Probe', 'Reamer', 'SlittingSaw', 'Tap', 'ThreadMill', 'Bullnose', 'VBit']`

### Traceback "dependency for shape not found"

Your tool library seems to be invalid in some cases, for example, drill_5.5mm.fctb references "drill.fcstd", which is not present in your shape folder. The traceback happened because the code assumed builtin shapes to be always available, as they are used as fallback when the shape was not found. To handle this smoothly for users, I now add built-in standard shapes to the folder if they are missing.

### Warning "ToolBit drill_1.0mm Parameter 'TipAngle' not found"

Your drill shapes, even when recognized as drills, do not fit the shape schema for a drill.
This brings up a fundamental issue: So far the assumption was: Drills, even if they are custom drills, MUST have some attributes like the TipAngle for CAM operations to work correctly. Some of your drills don't, hence they fail the validation and cause one of the tracebacks seen.

I *could* remove the TipAngle from the shape schema, but I think that some Ops have a "compensation" for the tip length, so your shape would still not work correctly in those. While I think the right thing to do in this case would be to fix your shape and add the missing TipAngle property, I think we need to allow for a transition period where even invalid specifications do not suddenly fail, given they likely were "good enough" for some types of operations.

So what I did to fix this:
- Shape schema violations are now no longer enforced, but the do still print a warning.
- I improved the error message - no traceback, just a friendlier warning.

### Tool add window does not load

This was a consequence of the errors above.

### Other improvements

- I changed the shape selector to show the shape ID on each button (in addition to the shape type)
- Custom shapes are now correctly sorted separately under "My Tools" in the selector dialog.

knipknap commented:
Fixed!

MisterMakerNL commented:
The ui look top notch!
But please look out that we have dark and light themes. So it should work in both!

knipknap commented:
Thanks for the feedback. I think that colors are managed in themes themselves; the code itself should not interfere with the colors. Hence if something were to be rendered incorrectly in one of the themes, I assume that the respective theme would need to be updated.

knipknap commented:
> That's not true for everything, we will get issues if you use images without backgrounds.

Ah, I see! I am not sure if I need to do anything specific to allow for styling the background. I used a [dedicated Qt widget](https://github.com/knipknap/FreeCAD/blob/btl/src/Mod/CAM/Path/Tool/shape/ui/shapewidget.py) wherever an image is used, so that it can be assigned a color.

The only other place is in a [cell in the list widget](https://github.com/knipknap/FreeCAD/blob/btl/src/Mod/CAM/Path/Tool/toolbit/ui/tablecell.py), but I assume these cells can be assigned colors as well.

> And black text without borders without backgrounds.

If I set any text colors, then I fetched them from the theme like this:

        palette = self.palette()
        bg_role = self.backgroundRole()
        bg_color = palette.color(bg_role)
        fg_role = self.foregroundRole()
        fg_color = palette.color(fg_role)

And that is pretty much only here:

https://github.com/knipknap/FreeCAD/blob/btl/src/Mod/CAM/Path/Tool/toolbit/ui/tablecell.py#L52

to style the subtitles in the cells. I am not sure if there is a safer way to do it. Without that, the tool summary would have the same color as the foreground text, I was trying to avoid that.

jffmichi commented:
OCL operations don't work with this PR:
```
01:21:58  Failed to identify tool for operation.
01:21:58  Surface.ERROR: Canceling 3D Surface operation. Error creating OCL cutter.
```
The issue seems to be SurfaceSupport.py (line 2437) expecting the tools to have a property "ShapeName" which I think you renamed to "ShapeID"?

jffmichi commented:
Do you still have the save button in the tool library on your mind? 😀

Feature request: would it be possible to remember the sorting mode in the tool dock?
Extra bonus: would it be possible to sort by label and handle numbers, e.g. sort "Drill 1.0mm", "Drill 6.0mm" and "Drill 10.0mm" nicely?

Thanks a lot for all your work

tarman3 commented:
I tried this branch
After open tool library see only tools from `/usr/lib/freecad/Mod/CAM/Tools`

Is it possible to import tool library which I used before?

knipknap commented:
@jffmichi Thanks a lot for testing and for tracking down the source of the problem in SurfaceSupport.py! All reported issues should now be fixed. I put your feature requests in my TODO, they will be done as part of the follow-up PR that I will place once this one here is merged.

@tarman3 Also thanks for reporting this (again). Error is identified and should now *really* be fixed as well :smile: .

tarman3 commented:
Checked
Thanks you

PS You can apply `pre-commit install` to exclude `[pre-commit.ci] auto fixes ...`

knipknap commented:
> PS You can apply `pre-commit install` to exclude `[pre-commit.ci] auto fixes ...` 

Ah, good tip, thanks! Done!

jffmichi commented:
Afaik the save button previous to this PR was doing exactly what the export button is doing now. I can happily live with it like this until you replace the dialog 👍

jffmichi commented:
Camotics tool export exports all tools as "Cylindrical":
```
  "69": {
    "units": "metric",
    "shape": "Cylindrical",
    "length": 50.0,
    "diameter": 4.0,
    "description": "Ball Nose End Mill 4.0mm 2F"
  },
```
Minor detail: I think camotics expects tool tables as "*.json". Currently we are exporting to *.ctbl.
![grafik](https://github.com/user-attachments/assets/b116fd41-462c-4cb2-b99d-29ea35ece003)

jffmichi commented:
LinuxCNC tool export outputs diameters with only 1 decimal, it would be great if you could change that to 3 (or use the user settings "Number of decimals"). E.g. a 1/8th inch endmill is exported as 3.2mm.

Minor detail: there are lonely "P"s in the resulting files. I don't have a LinuxCNC running here, maybe someone could check if this is imported correctly?

```
T6 P D3.2 ;End Mill 3.175mm 1F
```

knipknap commented:
Thanks again @jffmichi  - all issues/improvements done. I also improved the serializer tests to catch this in the future.

jffmichi commented:
Unfortunately the "Select a Tool Shape" dialog now only shows my custom tools, "Standard Tools" is empty. 

![grafik](https://github.com/user-attachments/assets/012ed835-63aa-464b-bcb6-0d997582b7ea)
![grafik](https://github.com/user-attachments/assets/ab48ad10-0fe3-48f2-bae3-55ea4c06683f)

knipknap commented:
The issue with the warnings in spite of having the correct body label should be fixed now.

The shape selector behavior was actually expected, because that is how the FreeCAD 1.0 behaves (even though BTL already adopted another approach). In FreeCAD 1.0, the expectation is that any shapes you use should be in your Shape folder. The "built-in tools" are only used as a tool folder template.

But I just changed that and now adopted the BTL behavior, I think it makes it easier.

jffmichi commented:
There seems to be a problem with the Chamfer shape. This is what the my 6mm chamfer looks like when opened with this PR (notice the TipDiameter attribute is missing):

![grafik](https://github.com/user-attachments/assets/d4a7bd6e-6ccf-44d8-90b7-de306aaf4a11)

This is what it looks like in FreeCAD 1.0.1 (how it should be):

![grafik](https://github.com/user-attachments/assets/8966454b-b4a2-4090-a46a-3fb4c97896c2)

knipknap commented:
@jffmichi  Thanks for your diligence in hunting down bugs :-). It is fixed and I tested, should work now.

jffmichi commented:
It seems im slowly running out of things to hunt... Thanks for all the fixes :smile:

sliptonic commented:
[tools.zip](https://github.com/user-attachments/files/20507155/tools.zip)

knipknap commented:
Oh man, that tool directory is a *mess* :laughing: . So much legacy and invalid data, a great test case!

Fixed/improvements pushed, everything should be handled gracefully now.

knipknap commented:
Dang, the tests fail because they are still on Python 2.10 :-(. I'll push a workaround...

sliptonic commented on src/Mod/CAM/Path/Tool/assets/ui/preferences.py:
```diff
@@ -0,0 +1,128 @@
+    def saveSettings(self):
+        # Check path is writable, then call Path.Preferences.setAssetPath()
+        asset_path = pathlib.Path(self.asset_path_edit.text())
+        param = FreeCAD.ParamGet("User parameter:BaseApp/Preferences/Path")
```
The correct parameter group should be BaseApp/Preferences/CAM

knipknap replied:
Good catch! Done

knipknap commented:
Great to hear, Thanks a lot! I can squash them or not, whatever you prefer.

Of course, if there are any bugs I will fix them, I am available in the next few weeks.

knipknap commented:
Awesome, Thanks to everyone who helped by testing, reviewing, and providing feedback!

tarman3 commented:
Good luck next week))

GeneGH commented:
This is a complete disaster in terms of roll-out.

The new scheme may be great, but there does not appear to be any documentation or facility to convert existing tools libraries.

The job isn't done until the paperwork is complete.
